stan/math/prim/scal/fun/grad_2F1.hpp

```diff
@@ -1,42 +1,109 @@
 #ifndef STAN_MATH_PRIM_SCAL_FUN_GRAD_2F1_HPP
 #define STAN_MATH_PRIM_SCAL_FUN_GRAD_2F1_HPP
 
+#include <stan/math/prim/scal/fun/sign.hpp>
+#include <stan/math/prim/scal/err/domain_error.hpp>
+#include <stan/math/prim/scal/err/check_2F1_converges.hpp>
 #include <cmath>
+#include <limits>
 
 namespace stan {
   namespace math {
 
-    // Gradient of the hypergeometric function 2F1(a, b | c | z)
-    // with respect to a and c
+    /**
+     * Gradients of the hypergeometric function, 2F1.
+     *
+     * Calculate the gradients of the hypergeometric function (2F1)
+     * as the power series stopping when the series converges
+     * to within <code>precision</code> or throwing when the
+     * function takes <code>max_steps</code> steps.
+     *
+     * This power-series representation converges for all gradients
+     * under the same conditions as the 2F1 function itself.
+     *
+     * @tparam T type of arguments and result
+     * @param[out] g_a1 g_a1 pointer to array of six values of type T, result.
+     * @param[out] g_b1 g_b1 pointer to array of six values of type T, result.
+     * @param[in] a1 a1 see generalized hypergeometric function definition.
+     * @param[in] a2 a2 see generalized hypergeometric function definition.
+     * @param[in] b1 b1 see generalized hypergeometric function definition.
+     * @param[in] z z see generalized hypergeometric function definition.
+     * @param[in] precision precision of the infinite sum. defaults to 1e-6
+     * @param[in] max_steps number of steps to take. defaults to 10,000
+     */
     template<typename T>
-    void grad_2F1(T& gradA, T& gradC, T a, T b, T c, T z, T precision = 1e-6) {
+    void grad_2F1(T& g_a1, T& g_b1, const T& a1, const T& a2, const T& b1,
+        const T& z, const T& precision = 1e-10, int max_steps = 1e5) {
+      check_2F1_converges("grad_2F1", a1, a2, b1, z);
+
+      using std::log;
       using std::fabs;
+      using std::exp;
+
+      g_a1 = 0.0;
+      g_b1 = 0.0;
+
+      T log_g_old[4];
+      for (T *q = log_g_old; q != log_g_old + 4; ++q)
```

Comment:
Done, skipped on the std::vector to stick to original usage of arrays, these functions could all be much clearer with std::vector for arguments and some returns. Went with the for loops on arrays and direct negation.